Open Judge System/Web/OJS.Web/Controllers/TempController.cs

```diff
@@ -368,5 +369,109 @@ public ActionResult MigrateSolutionSkeletons()
+                {
+                    this.Data.ContestCategories.Add(contestCategoryResponse.Data);
+                    this.Data.SaveChanges();
+                    scope.Complete();
+                    return this.Content($"The contest category is added!");
+                }
+            }
+            catch (Exception e)
+            {
+                return this.Content(
+                    $"Contest categories can't be migrated and exception {e}");
+            }
+        }
+
+        private async Task LoadContestCategoryAndAssignCheckerAndSubmissionTypes(ContestCategory contestCategory)
+        {
+            if (contestCategory == null)
+            {
+                return;
+            }
+
+            foreach (var contest in contestCategory.Contests)
+            {
+                foreach (var problemGroup in contest.ProblemGroups)
+                {
+                    this.LoadContestCategoryAndAssignCheckerAndSubmissionTypes(problemGroup.Problems);
+                }
+            }
+
+            var ids = contestCategory.Children.Select(x => x.Id).ToList();
+            contestCategory.Children.Clear();
+
+            foreach (var id in ids)
+            {
+                var contestCategoryResponse = await this.FetchContestCategory(id);
+
+                contestCategory.Children.Add(contestCategoryResponse.Data);
+                await this.LoadContestCategoryAndAssignCheckerAndSubmissionTypes(contestCategoryResponse.Data);
+            }
+        }
+
+        private async Task<ExternalDataRetrievalResult<ContestCategory>> FetchContestCategory(int id)
+        {
+            var contestCategoryResponse = await this.httpRequester.GetAsync<ContestCategory>(
+                new { id },
+                string.Format(UrlConstants.GetContestCategoryExportDataApiFormat, Settings.JudgeBaseUrl),
+                Settings.ApiKey);
+
+            if (!contestCategoryResponse.IsSuccess || contestCategoryResponse.Data == null)
+            {
+                throw new InvalidOperationException(contestCategoryResponse.ErrorMessage);
+            }
+
+            return contestCategoryResponse;
+        }
+
+        private void LoadContestCategoryAndAssignCheckerAndSubmissionTypes(IEnumerable<Problem> problems)
+        {
+            foreach (var problem in problems)
+            {
+                var checker = this.Data.Checkers.All().FirstOrDefault(x => x.Name == problem.Checker.Name);
+
+                problem.Checker = checker;
+
+                if (problem.ProblemSubmissionTypesSkeletons.Any())
+                {
+                    problem.SubmissionTypes = new List<SubmissionType>();
+
+                    foreach (var problemSubmissionType in problem.ProblemSubmissionTypesSkeletons)
+                    {
+                        var submissionType = this.Data.SubmissionTypes.All().FirstOrDefault(x => x.Name == problemSubmissionType.SubmissionType.Name);
```

Comment:
Here is the same. No need to call db from loop (worse within another loop). Could be passed like checkers as well.